stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -219,10 +219,34 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) =
     lmul!(D, copy_oftype(B, promote_op(*, eltype(B), eltype(D.diag))))
 
-(*)(A::AbstractMatrix, D::Diagonal) =
-    rmul!(copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))), D)
-(*)(D::Diagonal, A::AbstractMatrix) =
-    lmul!(D, copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))))
+# Using mul! instead of (l/r)mul! provides a performance boost by avoiding a copy.
+# This is not possible in general, but works if the elements are numbers
+# See #42321
+# Using runtime eltype checks instead of dispatch to avoid ambiguities
+function (*)(A::AbstractMatrix, D::Diagonal)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
+        nA, nD = size(A, 2), length(D.diag)
+        if nA != nD
+            throw(DimensionMismatch("second dimension of A, $nA, does not match the first of D, $nD"))
+        end
+        return mul!(similar(A, Tdest, size(A)), A, D)
+    end
+    return rmul!(copy_similar(A, Tdest), D)
+end
+function (*)(D::Diagonal, A::AbstractMatrix)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
+        nA, nD = size(A, 1), length(D.diag)
+        if nA != nD
+            throw(DimensionMismatch("second dimension of D, $nD, does not match the first of B, $nA"))
+        end
+        return mul!(similar(A, Tdest, size(A)), D, A)
+    end
+    return lmul!(D, copy_similar(A, Tdest))
+end
```

Comment:
It seems like the whole performance improvement can be achieved by replacing the above two methods by
```suggestion
(*)(A::AbstractMatrix, D::Diagonal) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), A, D)
(*)(D::Diagonal, A::AbstractMatrix) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), D, A)
```
In fact, the second isn't even necessary and could be removed(!), since the fallbacks handle everything. This avoids the second pass over the destination array.

Reply:
`+(::AbstractArray, ::Bool)` throw MethodError now.
So current `mul!` should be limited to Number Eltype.
We can define something like `*ₛ` to partially handle it:
```julia
function +ₛ end
Base.broadcasted(::typeof(+ₛ), ::Bool, bc) = bc
Base.broadcasted(::typeof(+ₛ), bc, ::Bool) = bc
Base.broadcasted(::typeof(+ₛ), bc₁, bc₂) = Base.broadcasted(+, bc₁, bc₂)
# what should we set in out for non-number eltype if alpha == beta == 0
Base.broadcasted(::typeof(+ₛ), ::Bool, ::Bool) = false 
```

Reply:
Where do you see that failure?

Reply:
I'm also not seeing extra allocations:
```julia
julia> @btime mul!($(similar(A)), $A, $DR);
  5.636 ms (2 allocations: 96 bytes)
```
We may have a few extra operations, but no extra allocations because `mul!` broadcasts right into the destination. Why can't we handle the alpha=beta=0 case? I assume this is all well tested, isn't it?

Reply:
```julia
julia> A = Diagonal([[1 2;3 4],[1 2;3 4]])
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [1 2; 3 4]      ⋅
     ⋅       [1 2; 3 4]

julia> B = [1,1]
2-element Vector{Int64}:
 1
 1

julia> A * B
2-element Vector{Matrix{Int64}}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> C =  A * B
2-element Vector{Matrix{Int64}}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> mul!(C,A,B)
ERROR: MethodError: no method matching +(::Matrix{Int64}, ::Bool)
```

Reply:
Ok, then I suggest we simply add a few 3-arg `mul!` methods:
```julia
@inline mul!(out::AbstractVector, A::Diagonal, in::AbstractVector) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::AbstractMatrix) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::Adjoint{<:Any,<:AbstractVecOrMat}) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::Transpose{<:Any,<:AbstractVecOrMat}) =
    out .= A.diag .* in

@inline mul!(out::AbstractMatrix, in::AbstractMatrix, A::Diagonal) =
    out .= in .* permutedims(A.diag)
@inline mul!(out::AbstractMatrix, in::Adjoint{<:Any,<:AbstractVecOrMat}, A::Diagonal) =
    out .= in .* permutedims(A.diag)
@inline mul!(out::AbstractMatrix, in::Transpose{<:Any,<:AbstractVecOrMat}, A::Diagonal) =
    out .= in .* permutedims(A.diag)
```

Reply:
This seems to be a good approach, and I'll admit the restriction to numbers didn't have any fundamental basis other than avoiding missing methods. I'll try adding the suggested methods in over the weekend and see if things work

Reply:
I have added the three element `mul!`, but this doesn't solve the issue with the five-element `mul!` that still remains.

```julia
julia> A = reshape([[1 2; 3 4], zeros(Int,2,2), zeros(Int, 2, 2), [5 6; 7 8]], 2, 2)
2×2 Matrix{Matrix{Int64}}:
 [1 2; 3 4]  [0 0; 0 0]
 [0 0; 0 0]  [5 6; 7 8]

julia> D = Diagonal(A)
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [1 2; 3 4]      ⋅     
     ⋅       [5 6; 7 8]

julia> mul!(similar(A), A, D, true, false)
ERROR: MethodError: no method matching +(::Matrix{Int64}, ::Bool)

julia> mul!(similar(A), A, D)
2×2 Matrix{Matrix{Int64}}:
 [7 10; 15 22]  [0 0; 0 0]
 [0 0; 0 0]     [67 78; 91 106]
```